src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,531 @@
+
+                var roleRequest = CreateRoleRequest(_ec2BaseUri, token);
+                var roleName = await GetHttpContent(roleRequest, "Failed to acquire EC2 role name.").ConfigureAwait(false);
+
+                var credentialsRequest = CreateCredentialsRequest(_ec2BaseUri, roleName, token);
+                var credentials = await GetHttpContent(credentialsRequest, "Failed to acquire EC2 credentials.").ConfigureAwait(false);
+
+                return credentials;
+            }
+
+            public static async Task<string> GetECSResponse(string relativeUri)
+            {
+                var credentialsRequest = new HttpRequestMessage
+                {
+                    RequestUri = new Uri(_ecsBaseUri, relativeUri),
+                    Method = HttpMethod.Get
+                };
+
+                return await GetHttpContent(credentialsRequest, "Failed to acquire ECS credentials.").ConfigureAwait(false);
+            }
+
+            // private static methods
+            private static HttpRequestMessage CreateCredentialsRequest(Uri baseUri, string roleName, string token)
+            {
+                var credentialsUri = new Uri(baseUri, "latest/meta-data/iam/security-credentials/");
+                var credentialsRequest = new HttpRequestMessage
+                {
+                    RequestUri = new Uri(credentialsUri, roleName),
+                    Method = HttpMethod.Get
+                };
+                credentialsRequest.Headers.Add("X-aws-ec2-metadata-token", token);
+
+                return credentialsRequest;
+            }
+
+            private static HttpRequestMessage CreateRoleRequest(Uri baseUri, string token)
+            {
+                var roleRequest = new HttpRequestMessage
+                {
+                    RequestUri = new Uri(baseUri, "latest/meta-data/iam/security-credentials/"),
+                    Method = HttpMethod.Get
+                };
+                roleRequest.Headers.Add("X-aws-ec2-metadata-token", token);
+
+                return roleRequest;
+            }
+
+            private static HttpRequestMessage CreateTokenRequest(Uri baseUri)
+            {
+                var tokenRequest = new HttpRequestMessage
+                {
+                    RequestUri = new Uri(baseUri, "latest/api/token"),
+                    Method = HttpMethod.Put,
+                };
+                tokenRequest.Headers.Add("X-aws-ec2-metadata-token-ttl-seconds", "30");
+
+                return tokenRequest;
+            }
+
+            private static async Task<string> GetHttpContent(HttpRequestMessage request, string exceptionMessage)
+            {
+                HttpResponseMessage response;
+                try
+                {
+                    response = await __httpClientInstance.Value.SendAsync(request).ConfigureAwait(false);
+                    response.EnsureSuccessStatusCode();
+                }
+                catch (OperationCanceledException ex)
```

Comment:
Is this exception even possible?

I don't see any CancellationToken being used.